File: apps/block-proxy/src/upstream/fastnear.rs
use std::time::Duration;

use bytes::Bytes;
use tracing::debug;

use crate::config::Config;

/// fastnear upstream fetcher.
///
/// Retrieves block data from the fastnear HTTP API (`mainnet.neardata.xyz` or
/// `testnet.neardata.xyz`). fastnear returns the canonical snake_case block format
/// so no transformation is needed — bytes are forwarded as-is.
pub struct FastnearUpstream {
    client: reqwest::Client,
    base_url: String,
    timeout: Duration,
}

impl FastnearUpstream {
    /// Creates a new `FastnearUpstream` from configuration.
    ///
    /// Builds a connection-pooled `reqwest::Client` (no client-level timeout —
    /// timeout is applied per-request per RESEARCH.md pattern 6).
    /// Maximum response body size (100 MB). Blocks can be multi-MB but
    /// anything larger than this indicates a misbehaving upstream.
    const MAX_BODY_SIZE: usize = 100 * 1024 * 1024;

    pub fn new(config: &Config) -> Self {
        let client = reqwest::Client::new();
        Self {
            client,
            base_url: config.fastnear_base_url(),
            timeout: Duration::from_secs(config.upstream_timeout_secs),
        }
    }

    /// Fetches a block by height from the fastnear API.
    ///
    /// URL: `{base_url}/v0/block/{height}`
    ///
    /// Per-request timeout is applied via `.timeout()` on the request builder.
    /// Returns raw response bytes without deserialization (fastnear format is canonical).
    pub async fn fetch(&self, height: u64) -> anyhow::Result<Bytes> {
        let url = format!("{}/v0/block/{}", self.base_url, height);
        let start = std::time::Instant::now();

        let response = self
            .client
            .get(&url)
            .timeout(self.timeout)
            .send()
            .await
            .map_err(|e| anyhow::anyhow!("fastnear request failed for block {}: {}", height, e))?;

        if response.status() == reqwest::StatusCode::NOT_FOUND {
            anyhow::bail!("block {} not found on fastnear", height);
        }

        let response = response
            .error_for_status()
            .map_err(|e| anyhow::anyhow!("fastnear returned error status for block {}: {}", height, e))?;

        let status = response.status();
        if let Some(len) = response.content_length() {
            if len as usize > Self::MAX_BODY_SIZE {
                anyhow::bail!("fastnear response too large for block {}: {} bytes", height, len);
            }
        }
        let data = response
            .bytes()
            .await
            .map_err(|e| anyhow::anyhow!("fastnear body read failed for block {}: {}", height, e))?;
        if data.len() > Self::MAX_BODY_SIZE {
            anyhow::bail!("fastnear response body too large for block {}: {} bytes", height, data.len());
        }

        let elapsed = start.elapsed();
        debug!(
            height,
            status = status.as_u16(),
            bytes = data.len(),
            latency_ms = elapsed.as_millis(),
            "fastnear upstream fetch complete"
        );

        Ok(data)
    }

    /// Fetches the latest finalized block from the fastnear API.
    ///
    /// URL: `{base_url}/v0/last_block/final`
    ///
    /// Used by the `/v0/last_block/final` endpoint handler. This result is NOT
    /// cached (per ENDP-02) — the handler calls this directly each time.
    pub async fn fetch_last_block_final(&self) -> anyhow::Result<Bytes> {
        let url = format!("{}/v0/last_block/final", self.base_url);
        let start = std::time::Instant::now();

        let response = self
            .client
            .get(&url)
            .timeout(self.timeout)
            .send()
            .await
            .map_err(|e| anyhow::anyhow!("fastnear last_block/final request failed: {}", e))?;

        if response.status() == reqwest::StatusCode::NOT_FOUND {
            anyhow::bail!("last_block/final not found on fastnear");
        }

        let response = response
            .error_for_status()
            .map_err(|e| anyhow::anyhow!("fastnear returned error status for last_block/final: {}", e))?;

        let status = response.status();
        if let Some(len) = response.content_length() {
            if len as usize > Self::MAX_BODY_SIZE {
                anyhow::bail!("fastnear response too large for last_block/final: {} bytes", len);
            }
        }
        let data = response
            .bytes()
            .await
            .map_err(|e| anyhow::anyhow!("fastnear body read failed for last_block/final: {}", e))?;
        if data.len() > Self::MAX_BODY_SIZE {
            anyhow::bail!("fastnear response body too large for last_block/final: {} bytes", data.len());
        }

        let elapsed = start.elapsed();
        debug!(
            status = status.as_u16(),
            bytes = data.len(),
            latency_ms = elapsed.as_millis(),
            "fastnear last_block/final fetch complete"
        );

        Ok(data)
    }
}

File: apps/block-proxy/src/upstream/fastnear.ts
import { logger } from 'nb-logger';

import type { UpstreamConfig } from '#config';
import * as metrics from '#metrics';
import type { UpstreamFetchError } from '#types';

const MAX_BODY_SIZE = 100 * 1024 * 1024; // 100 MB

/**
 * undici keeps a connection checked out until its body is read or cancelled.
 * Used where the body size is unknown; small 404 bodies are reclaimed anyway.
 */
const discardBody = async (response: Response): Promise<void> => {
  try {
    await response.body?.cancel();
  } catch {
    // Already consumed, or the socket is gone. Nothing left to release.
  }
};

/** `Retry-After` is either delta-seconds or an HTTP date. */
const parseRetryAfter = (header: null | string): number | undefined => {
  if (!header) return undefined;

  const seconds = Number(header);
  if (Number.isFinite(seconds) && seconds >= 0) return seconds * 1000;

  const date = Date.parse(header);
  if (!Number.isNaN(date)) return Math.max(0, date - Date.now());

  return undefined;
};

/**
 * One neardata-compatible endpoint. That is either FastNear itself or another
 * block-proxy instance, which serves the same two paths — so the only
 * difference is the base URL and whether an API key is attached.
 */
export class FastnearUpstream {
  private apiKey: string;

  private baseUrl: string;
  private timeoutMs: number;
  readonly name: string;

  constructor(upstream: UpstreamConfig, timeoutMs: number) {
    this.apiKey = upstream.apiKey;
    this.baseUrl = upstream.url;
    this.name = upstream.name;
    this.timeoutMs = timeoutMs;
  }

  private buildUrl(path: string): string {
    const url = `${this.baseUrl}${path}`;
    return this.apiKey
      ? `${url}?apiKey=${encodeURIComponent(this.apiKey)}`
      : url;
  }

  /** Records the 429 and tags the error so the pool can apply a cooldown. */
  private rateLimited(
    response: Response,
    start: number,
    height?: number,
  ): void {
    metrics.upstreamRateLimited.inc({ source: this.name });
    logger.warn(
      {
        height,
        latency_ms: Date.now() - start,
        retry_after: response.headers.get('retry-after'),
        source: this.name,
        status: 429,
      },
      'upstream rate limited',
    );
  }

  private statusError(response: Response, message: string): UpstreamFetchError {
    const err = new Error(message) as UpstreamFetchError;
    err.status = response.status;
    err.retryAfterMs = parseRetryAfter(response.headers.get('retry-after'));
    return err;
  }

  async fetch(height: number): Promise<Buffer> {
    const url = this.buildUrl(`/v0/block/${height}`);
    const start = Date.now();

    const response = await globalThis
      .fetch(url, { signal: AbortSignal.timeout(this.timeoutMs) })
      .catch((err) => {
        throw new Error(
          `${this.name} request failed for block ${height}: ${err}`,
        );
      });

    if (response.status === 404) {
      const err = new Error(
        `block ${height} not found on ${this.name}`,
      ) as UpstreamFetchError;
      err.notFound = true;
      err.status = 404;
      throw err;
    }

    if (response.status === 429) {
      this.rateLimited(response, start, height);
    }

    if (!response.ok) {
      await discardBody(response);
      throw this.statusError(
        response,
        `${this.name} returned error status for block ${height}: ${response.status}`,
      );
    }

    const contentLength = response.headers.get('content-length');
    if (contentLength && parseInt(contentLength, 10) > MAX_BODY_SIZE) {
      await discardBody(response);
      throw new Error(
        `${this.name} response too large for block ${height}: ${contentLength} bytes`,
      );
    }

    const arrayBuffer = await response.arrayBuffer();
    if (arrayBuffer.byteLength > MAX_BODY_SIZE) {
      throw new Error(
        `${this.name} response body too large for block ${height}: ${arrayBuffer.byteLength} bytes`,
      );
    }

    const data = Buffer.from(arrayBuffer);
    const elapsed = Date.now() - start;
    logger.debug(
      {
        bytes: data.length,
        height,
        latency_ms: elapsed,
        source: this.name,
        status: response.status,
      },
      'upstream fetch complete',
    );

    return data;
  }

  async fetchLastBlockFinal(): Promise<Buffer> {
    const url = this.buildUrl('/v0/last_block/final');
    const start = Date.now();

    const response = await globalThis
      .fetch(url, { signal: AbortSignal.timeout(this.timeoutMs) })
      .catch((err) => {
        throw new Error(`${this.name} last_block/final request failed: ${err}`);
      });

    if (response.status === 404) {
      const err = new Error(
        `last_block/final not found on ${this.name}`,
      ) as UpstreamFetchError;
      err.notFound = true;
      err.status = 404;
      throw err;
    }

    if (response.status === 429) {
      this.rateLimited(response, start);
    }

    if (!response.ok) {
      await discardBody(response);
      throw this.statusError(
        response,
        `${this.name} returned error status for last_block/final: ${response.status}`,
      );
    }

    const contentLength = response.headers.get('content-length');
    if (contentLength && parseInt(contentLength, 10) > MAX_BODY_SIZE) {
      await discardBody(response);
      throw new Error(
        `${this.name} last_block/final response too large: ${contentLength} bytes`,
      );
    }

    const arrayBuffer = await response.arrayBuffer();
    if (arrayBuffer.byteLength > MAX_BODY_SIZE) {
      throw new Error(
        `${this.name} last_block/final response body too large: ${arrayBuffer.byteLength} bytes`,
      );
    }

    const data = Buffer.from(arrayBuffer);
    logger.debug(
      {
        bytes: data.length,
        latency_ms: Date.now() - start,
        source: this.name,
        status: response.status,
      },
      'upstream last_block/final fetch complete',
    );

    return data;
  }
}
